### PR TITLE
Added a `Connection` class to isolate SQLRetry

### PR DESCRIPTION
**Pull request by EtienneBerubeShopify**
## Why?
Created an `Lhm::Connection` class to remove confusing logic and ensure that SQLRetry isn't passed around unnecessarily, rendering `@connection` useless. 

## How?
The `Lhm.setup(@connection)` creates an `Lhm::Connection` instance with the ActiveRecord connection. In order to make this change as non-breaking as possible, I thought of using mass-delegation instead of inheritance. The problem was that every method that is not "handled" by the new `Lhm::Connection` should be delegated to the ActiveRecord connection.

I had to improvise a way to do that, so I came up with `lib/connection.rb:9-22`. It's a bit convoluted but achieves what I was looking for quite well. I had to wrap it in a `begin-end` for internal ruby-specific reasons, but if anyone has any better idea, I am trying to make this **as clean and maintainable as possible**. 

TLDR (w.r.t `Lhm::Connection`); 
I get all the *relevant* instance methods used by ActiveRecord, remove the  "Object specific classes" (e.g. object_id, \_\_send\_\_, to_h, dup, etc.), remove the methods that should **NOT** be delegated (i.e. will contain some retry logic), then delegate them all to the `@connection` attribute.

Also, since now there's only one instance of `Lhm::SQLRetry`, the logging prefix was hard to manage. I whipped up a little something where I get the caller's file and `#camelise` it. It works really well and can be overridden by adding a log prefix explicitly.

## Focus
- Delegation process cleanliness
- I tried to be as "backwards compatible" as possible. So if you notice some edge-case or something that should now be deprecated, let me know.
- I had to change *some* logic in the test

### UPDATE:
The complex delegation has been changed to use `#method_missing` instead, as it drastically increases readability.

### CODE REVIEW

File: lib/lhm/connection.rb [hunk 1]
```diff
@@ -0,0 +1,57 @@
...
+
+    def select_value(query, retry_options = {})
+      exec_with_retries(:select_value, query, retry_options)
+    end
+
+    # Delegate ALL unknown method calls to ActiveRecord's connection.
+    # This ensures that there will be no breaking changes.
+    def method_missing(m, *args, &block)
+      @connection.public_send(m, *args, &block)
+    end
```
Comment (coding-chimp) [inline lib/lhm/connection.rb:31]: This would break in Ruby 3 if the underlying method has keyword arguments. There it would need to be either

```ruby
def method_missing(m, *args, **kwargs, &block)
  @connection.public_send(m, *args, **kwargs, &block)
end
```

or with the new argument forwarding operator 

```ruby
def method_missing(m, ...)
  @connection.public_send(m, ...)
end
```

However, the version with `**kwargs` is only supported since Ruby 2.7 and the one with `...` is supported since Ruby 3.0.

There are some ways around this, like defining the method differently based on the Ruby version

```ruby
if RUBY_VERSION < "3"
  def method_missing(m, *args, &block)
    @connection.public_send(m, *args, &block)
  end
else
  def method_missing(m, ...)
    @connection.public_send(m, ...)
  end
end
```

Alternatively, you could also use Ruby's [SimpleDelegator](https://docs.ruby-lang.org/en/3.0.0/SimpleDelegator.html) here. The class would need to subclass SimpleDelegator, you'd need to add a `super(connection)` to the end of your `initialize` and then you should be able to remove the `method_missing` definition. SimpleDelegator is doing more or less the same, but handles the different ruby versions for you.

Comment (EtienneBerubeShopify) [reply lib/lhm/connection.rb:31]: That's a GREAT catch! I'll investigate using a `SimpleDelegator` and see if it fits adequately in the codebase! If it works, I'll add appropriate comments to make sure it's not too confusing for the next contributor.

Comment (shuhaowu) [reply lib/lhm/connection.rb:31]: Do we need to use SimpleDelegator to only get around 1 if condition? I think 1 if condition is OK.

File: lib/lhm/chunker.rb [hunk 2]
```diff
@@ -78,7 +79,7 @@ def execute
     private
 
     def raise_on_non_pk_duplicate_warning
-      @connection.query("show warnings").each do |level, code, message|
+      @connection.execute("show warnings", @retry_options).each do |level, code, message|
```
Comment (shuhaowu) [inline lib/lhm/chunker.rb:82]: Is there a reason we changed from `query` to `execute`?

**Review by shuhaowu — APPROVED** [review]
Looks good. I would just do the `if RUBY VERSION` here, to avoid introducing yet another abstraction to replace a single line of code.
